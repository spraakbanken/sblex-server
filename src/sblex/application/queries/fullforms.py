import abc


class FullformQuery(abc.ABC):
    @abc.abstractmethod
    async def query(self, fragment: str) -> bytes:
        ...
