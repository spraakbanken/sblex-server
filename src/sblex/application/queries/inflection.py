import abc
from typing import TypedDict


class InflectionTableRow(TypedDict):
    form: str
    gf: str
    pos: str
    inhs: list[str]
    msd: str
    p: str


class InflectionTableQuery(abc.ABC):
    @abc.abstractmethod
    def query(self, paradigm: str, word: str) -> list[InflectionTableRow]: ...
