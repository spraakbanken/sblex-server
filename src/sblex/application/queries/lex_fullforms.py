import abc
from typing import TypedDict


class FullformLex(TypedDict):
    id: str
    fm: str
    fp: list[str]
    l: str  # noqa: E741
    gf: str
    p: str


class FullformLexQuery(abc.ABC):
    @abc.abstractmethod
    async def query(self, segment: str) -> list[FullformLex]: ...
