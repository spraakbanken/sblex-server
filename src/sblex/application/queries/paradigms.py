import abc
import typing


class Paradigms(abc.ABC):
    @abc.abstractmethod
    def query(self, s: str) -> typing.Tuple[str, list[str]]: ...

    def prepare_args(self, s: str) -> typing.Tuple[str, list[str]]:
        xs = [x.strip() for x in s.split(",") if len(x) > 0]
        baseform, pos = split_word_and_pos(xs[0])
        xs[0] = f"{baseform}:{pos}"
        return baseform, xs


def split_word_and_pos(s: str) -> typing.Tuple[str, str]:
    n = s.find(":")
    if n == -1:
        raise NoPartOfSpeechOnBaseform()
    word = s[:n].strip()
    pos = s[n + 1 :].strip()
    return word, pos


class NoPartOfSpeechOnBaseform(Exception):
    """The baseform must contain a Part-of-Speech tag."""
