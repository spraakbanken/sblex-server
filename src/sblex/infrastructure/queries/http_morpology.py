import logging
import sys

import httpx
from opentelemetry import trace
from sblex.fm import Morphology

logger = logging.getLogger(__name__)


class HttpMorphology(Morphology):
    def __init__(self, http_client: httpx.AsyncClient) -> None:
        self._http_client = http_client

    async def lookup(self, word: str, n: int = 0) -> bytes | None:
        with trace.get_tracer(__name__).start_as_current_span(
            sys._getframe().f_code.co_name
        ) as _process_api_span:
            response = await self._http_client.get(f"/morph/{word}/{n}")
            if response.status_code == 404:
                return None
            elif response.status_code >= 400:
                logger.error(
                    "Http lookup failed with status=%d",
                    response.status_code,
                    extra={"content": response.content},
                )
                return None
            return response.content

    async def lookup_from_bytes(self, s: bytes) -> bytes | None:
        response = await self._http_client.get(f"/morph/{s.decode('utf-8')}/0")
        return response.content
