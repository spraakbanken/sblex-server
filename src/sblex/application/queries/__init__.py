from sblex.application.queries.fullforms import FullformQuery
from sblex.application.queries.lex_fullforms import FullformLexQuery
from sblex.application.queries.lookup_lid import LookupLid

__all__ = ["FullformLexQuery", "LookupLid", "FullformQuery"]
