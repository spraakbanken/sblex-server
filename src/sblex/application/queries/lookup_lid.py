import abc
import logging
from typing import Any

from sblex.application.predicates import is_lemma, is_lexeme

logger = logging.getLogger(__name__)


class LookupLid(abc.ABC):
    @abc.abstractmethod
    def get_lemma(self, lid: str) -> dict[str, Any]:
        """Get lemma with given `lid`.

        Raises
        ------
        LemmaNotFound
            if lemma was not found
        LookupLidError
            custom error
        """
        ...

    @abc.abstractmethod
    def get_lexeme(self, lid: str) -> dict[str, Any]:
        """Get lexeme with given `lid`.

        Raises
        ------
        LexemeNotFound
            if lexeme was not found
        LookupLidError
            custom error
        """
        ...

    def get_by_lid(self, lid: str) -> dict[str, Any]:
        """Get lemma or lexeme with given `lid`.

        This method will analyze `lid` to determine if a lemma or lexeme is requested.

        Raises
        ------
        LemmaNotFound
            if lemma was not found
        LexemeNotFound
            if lexeme was not found
        LookupLidError
            custom error
        """
        if is_lemma(lid):
            logger.debug("calling LookupLid.get_lemma for '%s'", lid)
            return self.get_lemma(lid)
        elif is_lexeme(lid):
            logger.debug("calling LookupLid.get_lexeme for '%s'", lid)
            return self.get_lexeme(lid)
        else:
            return {}


class LookupLidError(Exception):
    """Raised when the SemanticRepository fails."""


class LemmaNotFound(LookupLidError, KeyError):
    """Raised when a Lemma is not found."""


class LexemeNotFound(LookupLidError, KeyError):
    """Raised when a Lexeme is not found."""
