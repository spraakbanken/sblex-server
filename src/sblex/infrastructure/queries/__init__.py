from sblex.infrastructure.queries.lookup_lex_fullforms import LookupFullformLexQuery
from sblex.infrastructure.queries.mem_lookup_lid import MemLookupLid

__all__ = ["MemLookupLid", "LookupFullformLexQuery"]
