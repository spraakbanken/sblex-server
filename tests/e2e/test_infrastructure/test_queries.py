import pytest
import pytest_asyncio
from httpx import AsyncClient
from sblex.infrastructure.queries.http_morpology import HttpMorphology


@pytest_asyncio.fixture(name="http_morphology")
async def fixture_http_morphology(fm_client: AsyncClient) -> HttpMorphology:
    return HttpMorphology(http_client=fm_client)


class TestHttpMorphology:
    @pytest.mark.asyncio
    async def test_lookup_from_bytes(self, http_morphology: HttpMorphology, snapshot) -> None:
        result = await http_morphology.lookup_from_bytes(b"dv\xc3\xa4ljes")

        assert result == snapshot
