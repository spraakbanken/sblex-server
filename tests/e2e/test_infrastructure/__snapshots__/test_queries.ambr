# serializer version: 1
# name: TestHttpMorphology.test_lookup_from_bytes
  b'[{"gf":"dv\xc3\xa4ljas","id":"dv\xc3\xa4ljas..vb.1","pos":"vb","is":[],"msd":"pres ind s-form","p":"vb_vs_dv\xc3\xa4ljas"},{"gf":"dv\xc3\xa4ljas","id":"dv\xc3\xa4ljas..vb.1","pos":"vb","is":[],"msd":"imper","p":"vb_vs_dv\xc3\xa4ljas"}]'
# ---
